ledger/participant-integration-api/src/main/scala/platform/store/completions/RangeCache.scala

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.completions
+
+import com.daml.ledger.participant.state.v1.Offset
+
+import scala.collection.SortedMap
+
+/** In memory cache implementation for completions. If size of cache exceeds maxItems, oldest elements will be removed,
+  * based on ledger offset
+  *
+  * @param range offset range of data stored in cache
+  * @param maxItems maximum amount of elements stored in cache
+  * @param cache map of cached elements
+  * @tparam T type of cached elements
+  */
+case class RangeCache[T] private (range: Range, maxItems: Int, cache: SortedMap[Offset, T]) {
+
+  assert(
+    !cache.lastOption.exists(_._1 > range.endInclusive) && !cache.headOption.exists(
+      _._1 <= range.startExclusive
+    ),
+    "Cached values cannot be outside cache range",
+  )
+
+  assert(cache.size <= maxItems, "Cache size cannot exceeds maxItems param")
+
+  /** Caches values. This method may remove oldest values (based on offset)
+    * @param appendRange range of proposed cache
+    * @param values proposed new values of cache
+    * @return new instance of cache containing conjunction of existing cache and proposed cache limited to maxItems
+    */
+  def append(
+      appendRange: Range,
+      values: SortedMap[Offset, T],
+  ): RangeCache[T] = {
+    if (range.areConsecutive(appendRange) || !range.areDisjointed(appendRange)) {
+      val allValues = values ++ cache
```

Comment:
Do you know what is the performance of concatenating two sortedMaps? I could not find the info in the usual place on scala-docs.

Reply:
I don't know. I would expect something around `(m + n)` where n is size of first and m size of second Map. It's complexity of merging 2 sorted lists.